### PR TITLE
Feature/achave11/551 native browser download

### PR DESCRIPTION
**Pull request by achave11**

### CODE REVIEW

File: test/service/test_dss_proxy.py [hunk 1]
```diff
@@ -34,17 +34,15 @@ def test_dss_files_proxy(self):
...
             azul_url = f'{self.base_url}/dss/files/{file_uuid}?replica=aws&version={file_version}'
-            before = time.time()
             response = requests.get(azul_url, allow_redirects=False)
-            request_duration = time.time() - before
-            self.assertLessEqual(retry_after, request_duration)
-            self.assertLess(request_duration, 2 * retry_after)
-            self.assertEqual(301, response.status_code)
-            self.assertEqual(azul_url + token, response.headers['Location'])
+            self.assertEqual(200, response.status_code)
+            json_response = json.loads(response.content)
```
Comment (hannes-ucsc) [inline test/service/test_dss_proxy.py]:
```suggestion
            body = json.loads(response.content)
```

Comment (hannes-ucsc) [reply test/service/test_dss_proxy.py]: Why is there no assertion on body['Retry-After'] and body['Status']?

I shouldn't have to catch these things. They drain my "mental energy" which is needed to solve bigger problems. Please try to be more meticulous.

File: CHANGELOG.yaml [hunk 2]
```diff
@@ -11,6 +11,16 @@ meta:
     - clients: the service API changed in a way that requires clients to be updated
 
 changes:
+  - title: Handle redirect response from DSS for processing in data-browser
```
Comment (hannes-ucsc) [inline CHANGELOG.yaml]:
```suggestion
  - title: Emulate HTTP redirects and retry-after in DSS file proxy 
```

File: CHANGELOG.yaml [hunk 3]
```diff
@@ -11,6 +11,16 @@ meta:
...
+    issues:
+    - https://github.com/DataBiosphere/azul/issues/551
+    upgrade:
+    - deploy
+    - client
+    notes: When a download request is made to the service from the data-browser, the service obtains a redirect response
```
Comment (hannes-ucsc) [inline CHANGELOG.yaml]: Document the behavior in the docstring of the request handling method. @Ben-Wu did a good job doing the same for the manifest endpoint in https://github.com/DataBiosphere/azul/pull/558/files#diff-a3dec4f6b35a08e7cb29086cb8766c3aR554 (plus my suggestions). Only put a reference to that docstring in these notes.

File: lambdas/service/app.py [hunk 4]
```diff
@@ -648,12 +649,44 @@ def generate_manifest(event, context):
...
+    `Retry-After` is the recommended number of seconds to wait before requesting the URL the `Location` field.
+    `Location` is the URL to make a GET request to in order to recheck the status.
+    If the client receives a response body with the `Status` field set to 301, the client should wait the number of
+    seconds specified in `Retry-After` and then request the URL given in the `Location` field. The URL will point
+    back at this endpoint so the client should expect a response of the same shape. Note that the actual HTTP
+    response is of status 200, only the `Status` field of the body will be 301. The intent is to emulate HTTP while
+    bypassing the default client behavior which, in most web browsers, is to ignore `Retry-After`. The response
+    described here is intended to be processed by client-side Javascript such that the recommended delay in
+    `Retry-After` can be handled in Javascript rather that relying on the native implementation by the web browser.
+    If the file checkout is done and the file is ready to be downloaded, the response will be:
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
    
    If the file checkout is done and the file is ready to be downloaded, the response will be:
```

File: lambdas/service/app.py [hunk 5]
```diff
@@ -685,6 +711,9 @@ def file_url(uuid, version):
 
 
 def _file_url(uuid, **kwargs):
-    return (config.service_endpoint() + '/dss/files' +
-            '/' + urllib.parse.quote(uuid, safe='') +
-            '?' + urllib.parse.urlencode(kwargs))
+    protocol = app.current_request.headers.get('x-forwarded-proto', 'http')
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Remind me to talk to you about refactoring thins.

File: test/service/test_dss_proxy.py [hunk 6]
```diff
@@ -21,30 +21,30 @@ def chalice_config(self):
...
+                                         url=dss_url,
+                                         status=301,
+                                         headers={'Location': dss_url + token,
+                                                  'Retry-After': str(retry_after)}))
+        azul_url = f'{self.base_url}/dss/files/{file_uuid}?replica=aws&version={file_version}'
+        response = requests.get(azul_url, allow_redirects=False)
+        self.assertEqual(200, response.status_code)
+        body = json.loads(response.content)
+        self.assertEqual(azul_url + token, body['Location'])
+        self.assertEqual(str(3), body["Retry-After"])
```
Comment (hannes-ucsc) [inline test/service/test_dss_proxy.py]: `Retry-After` should be an `int`.

File: test/service/test_response.py [hunk 7]
```diff
@@ -570,7 +570,7 @@ def test_transform_request_with_file_url(self):
             self.assertIn('url', file_data.keys())
             actual_url = urllib.parse.urlparse(file_data['url'])
             actual_query_vars = {k: one(v) for k, v in urllib.parse.parse_qs(actual_url.query).items()}
-            expected_base_url = urllib.parse.urlparse(config.service_endpoint())
+            expected_base_url = urllib.parse.urlparse(base_url)  # Should be identical to testing server URL
```
Comment (hannes-ucsc) [inline test/service/test_response.py]:
```suggestion
            expected_base_url = urllib.parse.urlparse(base_url)
```

File: lambdas/service/app.py (hunk 5 above)
Comment (achave11) [reply lambdas/service/app.py]: Done!

File: CHANGELOG.yaml [hunk 8]
```diff
@@ -11,6 +11,14 @@ meta:
     - clients: the service API changed in a way that requires clients to be updated
 
 changes:
+  - title: Emulate HTTP redirects and retry-after in DSS file proxy and endpoint change to `fetch/dss/files/{uuid}`
```
Comment (hannes-ucsc) [inline CHANGELOG.yaml]:
```suggestion
  - title: Emulate HTTP redirects and retry-after in DSS file proxy
```

You original version isn't even correct English.

File: CHANGELOG.yaml [hunk 9]
```diff
@@ -11,6 +11,14 @@ meta:
...
+    issues:
+    - https://github.com/DataBiosphere/azul/issues/551
+    upgrade:
+    - deploy
+    - client
+    notes: Refer to docstring for service API handler `/fetch/dss/files/{uuid}`
```
Comment (hannes-ucsc) [inline CHANGELOG.yaml]: Add 

> Note that the DSS file proxy endpoint changed from `…` to `…`

File: lambdas/service/app.py [hunk 10]
```diff
@@ -645,46 +643,90 @@ def generate_manifest(event, context):
...
-    of retries.
+    Initiate checking out file for download from the data store
 
-    This is only a best effort since we can only wait at most 30s before API Gateway times out.
+    parameters:
+        - name: uuid
+          in: path
+          type: string
+          description: UUID of the file to be checked out
+          - name: filters
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: This endpoint doesn't have a `filters` parameter.

File: lambdas/service/app.py [hunk 11]
```diff
@@ -645,46 +643,90 @@ def generate_manifest(event, context):
...
+
+
+@app.route(proxy_endpoint_path + '/{uuid}', methods=['GET'], cors=True)
 def files_proxy(uuid):
     """
-    Proxies requests to DSS' /files endpoint, with the intend of implementing Retry-After on the client's behalf. The
-    file URL that this service returns on the /repository/files endpoint points at this endpoint. This proxy endpoint
-    was added to prevent browsers that don't implement Retry-After from quickly exhausting their limit on the number
-    of retries.
+    Initiate checking out file for download from the data store
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: https://github.com/DataBiosphere/azul/pull/571#discussion_r239572882

I'm getting really frustrated with the apparent inability to just copy and paste my suggestions. This is not the first time that I gave a suggested change verbatim and it was botched even then.  You can also just click "Commit suggestion".

File: lambdas/service/app.py [hunk 12]
```diff
@@ -645,46 +643,90 @@ def generate_manifest(event, context):
...
+          in: path
+          type: string
+          description: UUID of the file to be checked out
+          - name: filters
+          in: path
+          type: string
+          description: Filters to be applied when requesting the file download
+
+    :return: A 200 response with a JSON body describing the status of the checkout from DSS.
+    The `uuid` parameter and `filters` are passed to the DSS in order to initiate checkout of the correct file
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Again, `filters` is not a parameter.

File: lambdas/service/app.py [hunk 13]
```diff
@@ -645,46 +643,90 @@ def generate_manifest(event, context):
...
+    in the correct replica. For more information refer to https://dss.data.humancellatlas.org under GET `/files/{uuid}`
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: This is the first time `replica` is mentioned. There is no mention that all query parameters are forwarded to the DSS even though we talked at length about this.

File: lambdas/service/app.py [hunk 14]
```diff
@@ -645,46 +643,90 @@ def generate_manifest(event, context):
...
-        return Response(body='', status_code=302, headers={'Location': location})
+        json_body = {"Status": 302, "Location": location}
+        return Response(body=json.dumps(json_body), status_code=200)
     else:
         dss_response.raise_for_status()
 
 
-def file_url(uuid, version):
-    return _file_url(uuid, version=version, replica='aws')
+def file_url(uuid, **kwargs):
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
def file_url(uuid, **params):
```

File: lambdas/service/app.py [hunk 15]
```diff
@@ -692,46 +690,100 @@ def generate_manifest(event, context):
...
+    bypassing the default client behavior which, in most web browsers, is to ignore `Retry-After`. The response
+    described here is intended to be processed by client-side Javascript such that the recommended delay in
+    `Retry-After` can be handled in Javascript rather that relying on the native implementation by the web browser.
+
+    If the file checkout is done and the file is ready to be downloaded, the response will be:
+
+    ```
+    {
+        "Status": 302,
+        "Location": "https://s3.file.url"
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]:
```suggestion
        "Location": "https://org-humancellatlas-dss-checkout.s3.amazonaws.com/blobs/…"
```

**Comment by hannes-ucsc** [comment]
I had to squash because https://github.com/DataBiosphere/azul/pull/571/commits/3e6827815af929dc1ee1d300bbd4140e12b2cb94 contained a [line diff](https://github.com/DataBiosphere/azul/commit/3e6827815af929dc1ee1d300bbd4140e12b2cb94#diff-5bd8b96cfbaa39a66bede725e170ad34R573) that looked like it didn't belong.

**Pull request merged**